src/stdlib/RandomMeasures.jl

```diff
@@ -211,42 +201,35 @@ function distribution(d::SizeBiasedSamplingProcess{<:PitmanYorProcess})
     return LocationScale(zero(d_rpm_d), d.surplus, dist)
 end
 
-function _logpdf_table(d::PitmanYorProcess, m::AbstractVector{Int})
+function _logpdf_table(d::PitmanYorProcess{T}, m::AbstractVector{Int}) where {T<:AbstractFloat}
     # compute the sum of all cluster counts
     sum_m = sum(m)
 
-    # shortcut if all cluster counts are zero
-    dd = d.d
-    T = typeof(dd)
-    iszero(sum_m) && return zeros(T, 1)
-
     # pre-calculations
     dθ = d.θ
-    z = log(sum_m + dθ)
+    z = log(sum_m + d.θ)
 
     # construct the table
-    K = length(m)
-    table = Vector{T}(undef, K)
-    contains_zero = false
-    @inbounds for i in 1:K
-        mi = m[i]
-
-        if iszero(mi)
-            if contains_zero
-                table[i] = -Inf
-            else
-                table[i] = log(dθ + dd * d.t) - z
-                contains_zero = true
-            end
-        else
-            table[i] = log(mi - dd) - z
-        end
+    has_zeros = any(i -> i == 0, m)
+    K = has_zeros ? length(m) : length(m)+1
+
+    # sanity check
+    @assert d.t == sum(i -> i>0, m)
+    table = ones(T, K) * T(-Inf)
+
+    if iszero(sum_m)
+        table[1] = T(0)
+        return table
     end
 
-    if !contains_zero
-        push!(table, log(dθ + dd * d.t) - z)
+    contains_zero = false
+    @inbounds for i in (has_zeros ? findall(j -> j>0, m) : 1:(K-1))
+        table[i] = T(log(m[i] - d.d) - z)
     end
 
+    k_new = has_zeros ? findfirst(j -> j == 0, m) : K
```

Comment:
As mentioned above, if `first_zero` is obtained at the beginning of this function this could be changed to
```suggestion
    k_new = first_zero === nothing ? K : first_zero
```